clippy_lints/src/hash_borrow_str_semantics.rs

```diff
@@ -0,0 +1,114 @@
+    /// `hash(x) == hash((x as Borrow<[u8]>).borrow()) == hash((x as Borrow<str>).borrow())`
+    ///
+    /// Unfortunately it doesn't hold as `hash("abc") != hash("abc".as_bytes())`.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use std::borrow::Borrow;
+    /// use std::hash::{Hash, Hasher};
+    ///
+    /// struct ExampleType {
+    ///     data: String
+    /// }
+    ///
+    /// impl Hash for ExampleType {
+    ///     fn hash<H: Hasher>(&self, state: &mut H) {
+    ///         self.data.hash(state);
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<str> for ExampleType {
+    ///     fn borrow(&self) -> &str {
+    ///         &self.data
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<[u8]> for ExampleType {
+    ///     fn borrow(&self) -> &[u8] {
+    ///         self.data.as_bytes()
+    ///     }
+    /// }
+    ///
+    /// fn hash(x: impl Hash) -> u64 {
+    ///    let mut hasher = std::collections::hash_map::DefaultHasher::new();
+    ///    x.hash(&mut hasher);
+    ///    hasher.finish()
+    /// }
+    ///
+    /// fn main() {
+    ///    let example = ExampleType { data: "abc".to_string() };
+    ///    assert_eq!(hash(&example), hash(example.borrow() as &str));
+    ///    assert_ne!(hash(&example), hash(example.borrow() as &[u8]));
+    /// }
+    ///
+    /// ```
+    ///
+    #[clippy::version = "1.75.0"]
+    pub HASH_BORROW_STR_SEMANTICS,
+    correctness,
+    "Ensures that the semantics of Borrow for Borrow<str> and Borrow<[u8]> are satisfied for Hash"
+}
+
+declare_lint_pass!(HashBorrowStrSemantics => [HASH_BORROW_STR_SEMANTICS]);
+
+impl LateLintPass<'_> for HashBorrowStrSemantics {
+    /// We are emitting this lint at the Hash impl of a type that implements all
+    /// three of Hash, Borrow<str> and Borrow<[u8]>.
+    ///
+    /// We check that we are in the Hash impl
+    /// Then we check that the type implements Borrow<str> and Borrow<[u8]>
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if let ItemKind::Impl(imp) = item.kind &&
+            let Some(TraitRef {path: Path {span, res, ..}, ..}) = imp.of_trait &&
+            let ty = cx.tcx.type_of(item.owner_id).instantiate_identity() &&
+            let Some(hash_id) = cx.tcx.get_diagnostic_item(sym::Hash) &&
+            Res::Def(DefKind::Trait, hash_id) == *res &&
+            let Some(borrow_id) = cx.tcx.get_diagnostic_item(sym::Borrow) &&
+            // since we are in the Hash impl, we don't need to check for that.
+            // we need only to check for Borrow<str> and Borrow<[u8]>
+            implements_trait(cx, ty, borrow_id, &[cx.tcx.types.str_.into()]) &&
+            implements_trait(cx, ty, borrow_id, &[Ty::new_slice(cx.tcx, cx.tcx.types.u8).into()])
+        {
+            span_lint_and_then(
+                cx,
+                HASH_BORROW_STR_SEMANTICS,
+                *span,
+                "can't satisfy the semantics of Borrow for Borrow<str> and Borrow<[u8]> for Hash",
```

Comment:
```suggestion
                "can't satisfy the semantics of `Hash` when both `Borrow<str>` and `Borrow<[u8]>` are implemented",
```
I feel like phrasing it like this makes it clearer, considering that this lint is primarily looking for the `Hash` impl and uses that span.

Reply:
I agree, but then `cargo dev fmt` pushes the line down in the ui test 😓

Reply:
>  I removed the line from the error check, I think it's fine to rely on just the ui test, right?

They aren't required yet, but I believe the idea is that soon it will be.

I assume this a problem because of the `//~^ ERROR:` pattern?
At least in rustc's test suite, IIRC, the annotation can be split up to multiple lines by using `//~|` on the next line and it inherits the level.

If that doesn't work, there's also the option to simply only include as many words of the error message as it fits in one line.
uitest only checks if the error message starts with the pattern.

Or, I guess, we can also just not specify it like you have it right now and make it the problem of whoever wants to make them required 😄 

@xFrednet do you know what the best solution to this problem is?

Reply:
Yes it's the //~^ pattern :D

Reply:
Interesting, no I don't know the best solution. I think, I would prefer the `//~^ ERROR:` for lints with a static message. We had a discussion, to maybe add an option to uitest, that allows us to specify the lint name instead of the message, but I believe this hasn't been implemented yet. Just having the `//~^ ERROR:` without a message SGTM :)

Reply:
I tried an empty `//~^ ERROR:` and it failed the bless check. Instead I used `//~^ ERROR: can't` and that seems to be okay.